src/Mvc/Mvc.ViewFeatures/src/RazorComponents/ComponentRenderer.cs

```diff
@@ -178,31 +181,16 @@ private IHtmlContent NonPrerenderedServerComponent(HttpContext context, ServerCo
 
         var currentInvocation = _serverComponentSerializer.SerializeInvocation(invocationId, type, parametersCollection, prerendered: false);
 
-        return new ComponentHtmlContent(ServerComponentSerializer.GetPreamble(currentInvocation));
+        var viewBuffer = new ViewBuffer(_viewBufferScope, nameof(ComponentRenderer), 16); // Preamble is fixed size
+        ServerComponentSerializer.AppendPreamble(viewBuffer, currentInvocation);
+        return viewBuffer;
     }
 
-    private static IHtmlContent NonPrerenderedWebAssemblyComponent(HttpContext context, Type type, ParameterView parametersCollection)
+    private IHtmlContent NonPrerenderedWebAssemblyComponent(HttpContext context, Type type, ParameterView parametersCollection)
     {
         var currentInvocation = WebAssemblyComponentSerializer.SerializeInvocation(type, parametersCollection, prerendered: false);
-
-        return new ComponentHtmlContent(WebAssemblyComponentSerializer.GetPreamble(currentInvocation));
-    }
-}
-
-internal class InvokedRenderModes
-{
-    public InvokedRenderModes(Mode mode)
-    {
-        Value = mode;
-    }
-
-    public Mode Value { get; set; }
-
-    internal enum Mode
-    {
-        None,
-        Server,
-        WebAssembly,
-        ServerAndWebAssembly
+        var viewBuffer = new ViewBuffer(_viewBufferScope, nameof(ComponentRenderer), 16); // Preamble is fixed size
```

Comment:
If we later change the preamble size, and don't realise we should also update this line, will we cause adverse perf effects? Is there any way to make it fail if we do that?

Reply:
Maybe the instantiation of the viewbuffer should be done inside `WebAssemblyComponentSerializer.AppendPreamble` so it's all localized to one bit of code.